### ISSUE TITLE
Feat: Type Engine - Convert Types to C

### ISSUE THREAD

**Issue by mahaloz** [angr/binsync#125]
Since we are using types between IDA and Binary Ninja more its time we start using a real TypeEngine as proposed in one of the internal BinSync meetings. 

## Proposal
In the core of BinSync, we need a `to_crepr` and `from_crepr` for every plugin that is supported in BinSync. In the BinSync database we will store everything as a `crepr`, then when a plugin pulls, we will detect which plugin it is and change the type as we return it to them. This will be actually not to hard to do. After this is done, it must be integrated into the diffing system. 

Deadline for this will be DEF CON Quals.

**Comment by arizvisa** [comment]
(super looking forward to see what y'all come up with, and how it might shape things in the future)

**Comment by arizvisa** [comment]
definitely looking forward to what y'all come up with.

**Comment by mahaloz** [comment]
## Idea Update
I've had some time to think about this and I think we should introduce this TypeEngine as a class that is called from some kind of event handler. 

First, we remove all the useless `push_*` functions and replace it with a single `push_artifact` function that is implemented inside the common controller. This means that the new `push_artifact` function will need to take an object of type `Artifact` in that function we do something like this:

```python
def push_artifact(artifact):
    artifact = self.push_event(artifact)
    if isinstance(artifact, StackVariable):
        ....
```

We check what the instance of the artifact is and we use that to get both the corresponding `set_*` function for the state and the correct `commit_msg` for that specific object. 

In the `push_event` function, we ask that each decompiler controller implement this event handler to correct the artifact in whatever way it needs to be corrected. I see this being in three ways:
1. Type, using their decompiler specific TypeEngine
2. StackVar Offset, using a function in their decompiler controller
3. Addr, using a rebase function in their decompiler controller

### Type Converter
Continuing with the old idea, we need the type converter to do some common things across decompilers:
1. Parse ctype strings, like `int [10];` to determine what the base type is 
2. Dump parsed ctype strings after modification

**Comment by mahaloz** [comment]
Closed by #163
